subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -524,19 +524,28 @@ public ServiceProvider getService(LookupContext context, TypeSpec serviceType) {
             if (providers == null) {
                 return null;
             }
-            List<ServiceProvider> candidates = new ArrayList<ServiceProvider>();
+            ServiceProvider singleCandidate = null;
+            List<ServiceProvider> candidates = null;
             for (Provider provider : providers) {
                 ServiceProvider service = provider.getService(context, serviceType);
                 if (service != null) {
-                    candidates.add(service);
+                    if (singleCandidate == null) {
+                        singleCandidate = service;
+                    } else {
+                        if (candidates == null) {
+                            candidates = new ArrayList<ServiceProvider>(2);
+                            candidates.add(singleCandidate);
+                        }
+                        candidates.add(service);
+                    }
                 }
             }
 
-            if (candidates.size() == 0) {
+            if (candidates == null && singleCandidate == null) {
```

Comment:
That's really interesting, so this method might be too long to inline more stuff into it or `ArrayList` is too long/complex to inline. I'd be curious if a `FastList` or some kind of transforming iterator would perform better while getting rid of the extra local variables and conditions. It would be cool if we could have reusable pieces that others can use in other parts of the code base instead of copying this code into other hotspots. Just some food for thought really, nothing blocking this PR :)